### PR TITLE
[Feature] Support list partition in be & add/drop list partition

### LINKED ISSUE #4035: Support list partition【Moderate】
# Requirement Description
1. List partition design.
2. List partition implementation, including table creation, loading and query.

```
CREATE TABLE h2 (
  c1 INT,
  c2 INT
)
PARTITION BY LIST(c1) (
  PARTITION p0 VALUES IN (1, 4, 7),
  PARTITION p1 VALUES IN (2, 5, 8)
)
DISTRIBUTED BY HASH(c2) BUCKETS 3;
```

# Mentor
@wyb 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #4035

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 7 / 15 (46.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|15|46.67%|[769, 770, 771, 772, 773, 774, 775, 777]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 8 / 20 (40.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|0|4|00.00%|[3058, 3059, 3060, 3061]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|7|15|46.67%|[769, 770, 771, 772, 773, 774, 775, 777]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 22 / 39 (56.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|0|4|00.00%|[3058, 3059, 3060, 3061]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|21|34|61.76%|[769, 770, 771, 772, 773, 774, 775, 777, 857, 858, 860, 861, 865]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|[]|

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -315,17 +336,18 @@ Status OlapTablePartitionParam::find_tablets(Chunk* chunk, std::vector<OlapTable
         ChunkRow row;
         row.columns = &partition_columns;
         row.index = 0;
+        bool is_by_value_partition = _t_param.partitions[0].__isset.in_keys;
```
Comment (wyb) [inline be/src/exec/tablet_info.cpp:339]: is_list_partition?
as a member variable?

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 2]
```diff
@@ -826,24 +836,34 @@ private void dropPartition(long dbId, String partitionName, boolean isForceDrop,
...
-                        isLakeTable());
-            } else if (!reserveTablets) {
-                GlobalStateMgr.getCurrentState().onErasePartition(partition);
-            }
+                // drop partition info
+                rangePartitionInfo.dropPartition(partition.getId());
+            } else if (partitionInfo.getType() == PartitionType.LIST) {
+                ListPartitionInfo listPartitionInfo = (ListPartitionInfo) partitionInfo;
+                if (!isForceDrop) {
+                    // TODO: recycle list partition
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:859]: report an error?

**Pull request closed without merging**